File: agent.py
import asyncio
import json
import logging

import aiohttp
from pydantic import ValidationError

from config import (
    CONFIDENT_API_KEY,
    CONFIDENT_WS_BASE_URL,
    HEARTBEAT_INTERVAL_S,
    INITIAL_RECONNECT_DELAY_S,
    MAX_RECONNECT_DELAY_S,
)
from schemas import RelayRequest, RelayResponse, RelayResponseType, ResponseMode

logger = logging.getLogger("relay-agent")


class RelayAgent:
    def __init__(
        self,
        server_url: str,
        api_key: str,
    ):
        self.server_url = server_url
        self.api_key = api_key
        self._reconnect_delay = INITIAL_RECONNECT_DELAY_S
        self._running = True

    async def start(self):
        while self._running:
            try:
                await self._connect()
            except Exception as e:
                logger.error(f"connection error: {e}")
            if not self._running:
                break
            logger.info(f"reconnecting in {self._reconnect_delay}s")
            await asyncio.sleep(self._reconnect_delay)
            self._reconnect_delay = min(
                self._reconnect_delay * 2, MAX_RECONNECT_DELAY_S
            )

    async def _connect(self):
        headers = {
            "Authorization": f"Bearer {self.api_key}"
        }
        async with aiohttp.ClientSession() as session:
            async with session.ws_connect(
                self.server_url, 
                headers=headers, 
                heartbeat=HEARTBEAT_INTERVAL_S
            ) as ws:
                logger.info(f"connected to {self.server_url}")
                self._reconnect_delay = INITIAL_RECONNECT_DELAY_S
                await self._listen(ws, session)

    async def _listen(
        self, 
        ws: aiohttp.ClientWebSocketResponse, 
        session: aiohttp.ClientSession
    ):
        async for msg in ws:
            if msg.type == aiohttp.WSMsgType.TEXT:
                asyncio.create_task(self._handle_request(ws, session, msg.data))
            elif msg.type in (
                aiohttp.WSMsgType.CLOSED,
                aiohttp.WSMsgType.ERROR,
            ):
                logger.info(f"WebSocket closed or errored: msg.type={msg.type}, msg={msg}")
                break

    async def _handle_request(
        self,
        ws: aiohttp.ClientWebSocketResponse,
        session: aiohttp.ClientSession,
        raw: str,
    ):
        try:
            request = RelayRequest.model_validate_json(raw)
        except ValidationError as e:
            logger.error(f"failed to parse relay request: {e}")
            return

        try:
            timeout = aiohttp.ClientTimeout(total=request.timeout)
            kwargs: dict = {
                "method": request.method,
                "url": request.url,
                "headers": request.headers,
                "timeout": timeout,
            }
            if request.method.upper() != "GET" and request.body is not None:
                kwargs["json"] = request.body

            logger.info(f"[relay] {request.method} {request.url} (request {request.id})")
            async with session.request(**kwargs) as response:
                is_streaming = request.response_mode in (
                    ResponseMode.SSE_STREAMING,
                    ResponseMode.HTTP_STREAMING,
                )

                if is_streaming:
                    # Send response headers first
                    header_msg = RelayResponse(
                        id=request.id,
                        type=RelayResponseType.CHUNK,
                        statusCode=response.status,
                        headers=dict(response.headers),
                        body="",
                    )
                    await ws.send_json(header_msg.model_dump())

                    # Stream chunks as they arrive
                    async for raw_chunk in response.content.iter_any():
                        text = raw_chunk.decode("utf-8", errors="replace")
                        if not text:
                            continue
                        chunk_msg = RelayResponse(
                            id=request.id,
                            type=RelayResponseType.CHUNK,
                            statusCode=response.status,
                            headers={},
                            body=text,
                        )
                        await ws.send_json(chunk_msg.model_dump())

                    # Signal stream end
                    end_msg = RelayResponse(
                        id=request.id,
                        type=RelayResponseType.STREAM_END,
                        statusCode=response.status,
                        headers={},
                        body="",
                    )
                    await ws.send_json(end_msg.model_dump())
                    logger.info(f"[relay] {request.method} {request.url} : {response.status} streamed (request {request.id})")
                else:
                    response_body = await response.text()
                    logger.info(f"[relay] {request.method} {request.url} : {response.status} (request {request.id})")

                    relay_response = RelayResponse(
                        id=request.id,
                        type=RelayResponseType.RESPONSE,
                        statusCode=response.status,
                        headers=dict(response.headers),
                        body=response_body,
                    )
                    await ws.send_json(relay_response.model_dump())
        except Exception as e:
            logger.error(f"request {request.id} failed: {e}")
            error_response = RelayResponse(
                id=request.id,
                type=RelayResponseType.ERROR,
                statusCode=502,
                headers={},
                body=json.dumps({"error": str(e)}),
            )
            await ws.send_json(error_response.model_dump())

    def stop(self):
        self._running = False


def create_agent() -> RelayAgent:
    if not CONFIDENT_API_KEY:
        raise ValueError("CONFIDENT_API_KEY is required")
    if not CONFIDENT_WS_BASE_URL:
        raise ValueError("CONFIDENT_WS_BASE_URL is required")

    return RelayAgent(
        server_url=CONFIDENT_WS_BASE_URL,
        api_key=CONFIDENT_API_KEY,
    )

File: go/agent.go
package confidentagent

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"os/signal"
	"strings"
	"sync"
	"sync/atomic"
	"syscall"
	"time"

	"github.com/gorilla/websocket"
)

func logInfo(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "%s [relay-agent] INFO: %s\n", time.Now().Format("2006-01-02 15:04:05,000"), fmt.Sprintf(format, args...))
}

func logError(format string, args ...any) {
	fmt.Fprintf(os.Stderr, "%s [relay-agent] ERROR: %s\n", time.Now().Format("2006-01-02 15:04:05,000"), fmt.Sprintf(format, args...))
}

type RelayAgent struct {
	ServerURL string
	APIKey    string
	HandlerFn HandlerFunc

	reconnectDelay int
	running        atomic.Bool
	mu             sync.Mutex
	writeMu        sync.Mutex
	conn           *websocket.Conn
}

func NewRelayAgent(serverURL string, apiKey string, handlerFn HandlerFunc) *RelayAgent {
	agent := &RelayAgent{
		ServerURL:      serverURL,
		APIKey:         apiKey,
		HandlerFn:      handlerFn,
		reconnectDelay: InitialReconnectDelayS,
	}
	agent.running.Store(true)
	return agent
}

func (a *RelayAgent) Start(ctx context.Context) error {
	for a.running.Load() {
		if err := a.connect(ctx); err != nil {
			logError("connection error: %v", err)
		}
		if !a.running.Load() || ctx.Err() != nil {
			break
		}
		logInfo("reconnecting in %ds", a.reconnectDelay)
		select {
		case <-ctx.Done():
			return nil
		case <-time.After(time.Duration(a.reconnectDelay) * time.Second):
		}
		a.reconnectDelay = min(a.reconnectDelay*2, MaxReconnectDelayS)
	}
	return nil
}

func (a *RelayAgent) connect(ctx context.Context) error {
	header := http.Header{"Authorization": {"Bearer " + a.APIKey}}
	conn, response, err := websocket.DefaultDialer.DialContext(ctx, a.ServerURL, header)
	if err != nil {
		if response != nil {
			response.Body.Close()
		}
		return err
	}
	defer conn.Close()

	a.mu.Lock()
	a.conn = conn
	a.mu.Unlock()

	logInfo("connected to %s", a.ServerURL)
	a.reconnectDelay = InitialReconnectDelayS

	return a.listen(ctx, conn)
}

func (a *RelayAgent) listen(ctx context.Context, conn *websocket.Conn) error {
	readWait := 2 * HeartbeatIntervalS * time.Second
	conn.SetReadDeadline(time.Now().Add(readWait))
	conn.SetPongHandler(func(string) error {
		return conn.SetReadDeadline(time.Now().Add(readWait))
	})

	done := make(chan struct{})
	defer close(done)

	go func() {
		ticker := time.NewTicker(HeartbeatIntervalS * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-done:
				return
			case <-ctx.Done():
				conn.Close()
				return
			case <-ticker.C:
				if err := conn.WriteControl(websocket.PingMessage, nil, time.Now().Add(10*time.Second)); err != nil {
					return
				}
			}
		}
	}()

	for {
		messageType, raw, err := conn.ReadMessage()
		if err != nil {
			return err
		}
		if messageType == websocket.TextMessage {
			go a.handleRequest(ctx, conn, raw)
		}
	}
}

func (a *RelayAgent) handleRequest(ctx context.Context, conn *websocket.Conn, raw []byte) {
	var request RelayRequest
	if err := json.Unmarshal(raw, &request); err != nil {
		logError("failed to parse relay request: %v", err)
		return
	}

	if a.HandlerFn != nil {
		logInfo("[handler] running handler (request %s)", request.ID)
		relayResponse := executeHandler(a.HandlerFn, request)
		a.sendJSON(conn, relayResponse)
		logInfo("[handler] %d (request %s)", relayResponse.StatusCode, request.ID)
		return
	}

	var bodyReader io.Reader
	hasBody := strings.ToUpper(request.Method) != "GET" && request.Body != nil
	if hasBody {
		payload, err := json.Marshal(request.Body)
		if err != nil {
			a.sendRequestError(conn, request.ID, err)
			return
		}
		bodyReader = bytes.NewReader(payload)
	}

	requestCtx, cancel := context.WithTimeout(ctx, time.Duration(request.Timeout)*time.Second)
	defer cancel()

	httpRequest, err := http.NewRequestWithContext(requestCtx, request.Method, request.URL, bodyReader)
	if err != nil {
		a.sendRequestError(conn, request.ID, err)
		return
	}
	if hasBody {
		httpRequest.Header.Set("Content-Type", "application/json")
	}
	for key, value := range request.Headers {
		httpRequest.Header.Set(key, value)
	}

	logInfo("[relay] %s %s (request %s)", request.Method, request.URL, request.ID)
	response, err := http.DefaultClient.Do(httpRequest)
	if err != nil {
		a.sendRequestError(conn, request.ID, err)
		return
	}
	defer response.Body.Close()

	isStreaming := request.ResponseMode == ResponseModeSSEStreaming ||
		request.ResponseMode == ResponseModeHTTPStreaming

	if isStreaming {
		// Send response headers first
		a.sendJSON(conn, RelayResponse{
			ID:         request.ID,
			Type:       RelayResponseTypeChunk,
			StatusCode: response.StatusCode,
			Headers:    flattenHeaders(response.Header),
			Body:       "",
		})

		// Stream chunks as they arrive
		reader := bufio.NewReader(response.Body)
		buffer := make([]byte, 4096)
		for {
			n, readErr := reader.Read(buffer)
			if n > 0 {
				a.sendJSON(conn, RelayResponse{
					ID:         request.ID,
					Type:       RelayResponseTypeChunk,
					StatusCode: response.StatusCode,
					Headers:    map[string]string{},
					Body:       string(buffer[:n]),
				})
			}
			if readErr != nil {
				if !errors.Is(readErr, io.EOF) {
					a.sendRequestError(conn, request.ID, readErr)
					return
				}
				break
			}
		}

		// Signal stream end
		a.sendJSON(conn, RelayResponse{
			ID:         request.ID,
			Type:       RelayResponseTypeStreamEnd,
			StatusCode: response.StatusCode,
			Headers:    map[string]string{},
			Body:       "",
		})
		logInfo("[relay] %s %s : %d streamed (request %s)", request.Method, request.URL, response.StatusCode, request.ID)
	} else {
		responseBody, readErr := io.ReadAll(response.Body)
		if readErr != nil {
			a.sendRequestError(conn, request.ID, readErr)
			return
		}
		logInfo("[relay] %s %s : %d (request %s)", request.Method, request.URL, response.StatusCode, request.ID)

		a.sendJSON(conn, RelayResponse{
			ID:         request.ID,
			Type:       RelayResponseTypeResponse,
			StatusCode: response.StatusCode,
			Headers:    flattenHeaders(response.Header),
			Body:       string(responseBody),
		})
	}
}

func (a *RelayAgent) sendJSON(conn *websocket.Conn, response RelayResponse) {
	payload, err := json.Marshal(response)
	if err != nil {
		logError("failed to serialize relay response for request %s: %v", response.ID, err)
		return
	}
	a.writeMu.Lock()
	defer a.writeMu.Unlock()
	if err := conn.WriteMessage(websocket.TextMessage, payload); err != nil {
		logError("failed to send relay response for request %s: %v", response.ID, err)
	}
}

func (a *RelayAgent) sendRequestError(conn *websocket.Conn, requestID string, err error) {
	logError("request %s failed: %v", requestID, err)
	a.sendJSON(conn, RelayResponse{
		ID:         requestID,
		Type:       RelayResponseTypeError,
		StatusCode: 502,
		Headers:    map[string]string{},
		Body:       marshalError(err.Error()),
	})
}

func flattenHeaders(header http.Header) map[string]string {
	headers := map[string]string{}
	for key, values := range header {
		if len(values) > 0 {
			headers[key] = values[0]
		}
	}
	return headers
}

func (a *RelayAgent) Stop() {
	a.running.Store(false)
	a.mu.Lock()
	if a.conn != nil {
		a.conn.Close()
	}
	a.mu.Unlock()
}

func CreateAgent() (*RelayAgent, error) {
	if confidentAPIKey() == "" {
		return nil, errors.New("CONFIDENT_API_KEY is required")
	}
	if confidentWSBaseURL() == "" {
		return nil, errors.New("CONFIDENT_WS_BASE_URL is required")
	}

	var handlerFn HandlerFunc
	if len(registered) > 1 {
		return nil, &HandlerError{Message: fmt.Sprintf(
			"multiple handler functions registered (%d) — exactly one is allowed per agent",
			len(registered),
		)}
	}
	if len(registered) == 1 {
		handlerFn = registered[0]
		logInfo("handler mode: using registered handler")
	}

	return NewRelayAgent(confidentWSBaseURL(), confidentAPIKey(), handlerFn), nil
}

func Run() error {
	agent, err := CreateAgent()
	if err != nil {
		return err
	}

	ctx, stop := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
	defer stop()
	go func() {
		<-ctx.Done()
		agent.Stop()
	}()

	logInfo("starting relay agent")
	err = agent.Start(ctx)
	logInfo("relay agent stopped")
	return err
}
